### PR TITLE
Cyclopedia Player tab (WIP)

### PR DESCRIPTION
**Pull request by ArturKnopik**
Player Cyclopedia
image

### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

Todo: fix god cap (Argument 2 has out-of-range value for unsigned int: 4294967395)
fix magic level percentage

### CODE REVIEW

File: data/scripts/network/cyclopediaPlayerInfo.lua [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+	msg:addU16(100 * Game.getExperienceStage(player:getLevel())) -- base exp rate, should exp rate 
+	msg:addU32(0); -- turnament exp (deprecated) 
+	msg:addU16(0); -- bonus for lov level
+	msg:addU16(0) -- XPBoost
+	msg:addU16(100) -- 100% bonus of base exp bonus 
+	msg:addU16(100) -- stamina boost(100=x1.0)
+	msg:addByte(0x00) -- can buy exp boost
+
+
```
Comment (omarcopires) [inline data/scripts/network/cyclopediaPlayerInfo.lua:81]: double spaces

File: data/scripts/network/cyclopediaPlayerInfo.lua [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	msg:addU16(player:getMana())
+	msg:addU16(player:getMaxMana())
+	msg:addByte(player:getSoul())
+	msg:addU16(player:getStamina())
+
+    local foodTime = 0
+	local condition = player:getCondition(CONDITION_REGENERATION, CONDITIONID_DEFAULT) 
+    if condition then
+        foodTime = condition:getTicks() /  1000
+        print(foodTime)
```
Comment (omarcopires) [inline data/scripts/network/cyclopediaPlayerInfo.lua:93]: print

File: data/scripts/network/cyclopediaPlayerInfo.lua [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    end
+	msg:addU16(foodTime) --health tick??
+	msg:addU16(player:getOfflineTrainingTime() / 60 / 1000)
+	msg:addU16(player:getSpeed() / 2)
+	msg:addU16(player:getBaseSpeed() / 2)
+	msg:addU32(player:getCapacity()) -- cap + boost
+	msg:addU32(player:getCapacity())
+	msg:addU32(player:getFreeCapacity())
+
+	msg:addByte(8) -- ??
```
Comment (omarcopires) [inline data/scripts/network/cyclopediaPlayerInfo.lua:103]: blessings?

File: data/scripts/network/cyclopediaPlayerInfo.lua (hunk 1 above)
Comment (ArturKnopik) [reply data/scripts/network/cyclopediaPlayerInfo.lua:81]: i know, work in progress

File: data/scripts/network/cyclopediaPlayerInfo.lua (hunk 3 above)
Comment (ArturKnopik) [reply data/scripts/network/cyclopediaPlayerInfo.lua:103]: i thing it's skills size, mlvl + basic skills

**Comment by MillhioreBT** [comment]
Ok I've been checking it...
My recommendation is that you have to move the send methods to where they belong to the player. (`data/lib/core/player.lua`)

the name of the file in network, it should be called, `cyclopedia_character.lua` since that is the name of the tab from the cyclopedia that it manages

In addition, I think the code is better this way, each method in a related table so as not to have to do if if if if, and also it could be merged without the need to declare even the missing functions, they can be added in other PRs, in case of emergency:

<details>

<summary>cyclopedia_character.lua</summary>

```lua
--[[
    Not all methods are written in the library, it's just a matter of writing them and they will be automatically used.

    The methods should be written in data/lib/core/player.lua.
]]--

local BASIC_INFO = 0x00
local GENERAL_STATS = 0x01
local COMBAT_STATS = 0x02
local RECENT_DEATHS = 0x03
local RECENT_PVP_KILLS = 0x04
local ACHIEVEMENTS = 0x05
local ITEM_SUMMARY = 0x06
local APPEARANCES = 0x07
local STORE = 0x08
local INSPECTION = 0x09
local BADGES = 0x0A
local TITLES = 0x0B

local methods = {
    [BASIC_INFO] = "sendBasicInfo",
    [GENERAL_STATS] = "sendGeneralStats",
    [COMBAT_STATS] = "sendCombatStats",
    [RECENT_DEATHS] = "sendRecentDeaths",
    [RECENT_PVP_KILLS] = "sendRecentPvpKills",
    [ACHIEVEMENTS] = "sendAchievements",
    [ITEM_SUMMARY] = "sendItemSummary",
    [APPEARANCES] = "sendAppearances",
    [STORE] = "sendStore",
    [INSPECTION] = "sendInspection",
    [BADGES] = "sendBadges",
    [TITLES] = "sendTitles"
}

local handler = PacketHandler(0xE5)

function handler.onReceive(player, msg)
    msg:skipBytes(4)
    local type = msg:getByte()
    local method = Player[methods[type]]
    if method then
        local response = NetworkMessage()
        response:addByte(0xDA)
        response:addByte(type)
        response:addByte(0x00)
        method(player, response)
    end
end

handler:register()
```

</details>

Here is a preview of the functions in player.lua
If you want you can add the rest using this same format or just leave these...
<details>

<summary>player.lua</summary>

```lua
function Player.getBlessings(self)
	local blessings = 0
	for i = 1, 8 do
		if self:hasBlessing(i) then
			blessings = blessings + 1
		end
	end
	return blessings
end

function Player.sendBasicInfo(self, msg)
	msg:addString(self:getName())
	msg:addString(self:getVocation():getName())
	msg:addU16(self:getLevel())

	local outfit = self:getOutfit()
	if outfit.lookType ~= 0 then
		msg:addU16(outfit.lookType)
		msg:addByte(outfit.lookHead)
		msg:addByte(outfit.lookBody)
		msg:addByte(outfit.lookLegs)
		msg:addByte(outfit.lookFeet)
		msg:addByte(outfit.lookAddons)
	else
		msg:addU16(outfit.lookTypeEx)
	end

	msg:addByte(0) -- hide stamina
	msg:addByte(0) -- enable store summary & character titles
	msg:addString("") -- character title

	msg:sendToPlayer(self)
	msg:delete()
	return true
end

-- hardcoded in client
local skillsId = {11, 9, 8, 10, 7, 6, 13}

local function getLevelPercent(player)
	local currentLevelExp = Game.getExperienceForLevel(player:getLevel())
	local nextLvlExp = Game.getExperienceForLevel(player:getLevel() +1)
	local playerExp = player:getExperience()
	return (playerExp - currentLevelExp) * 100 / (nextLvlExp - currentLevelExp)
end

function Player.sendGeneralStats(self, msg)
	msg:addU64(self:getExperience())
	msg:addU16(self:getLevel())
	msg:addByte(math.floor(getLevelPercent(self))) -- level percent

	msg:addU16(Game.getExperienceStage(self:getLevel()) * 100) -- base exp rate, should exp rate 
	msg:addU32(0) -- turnament exp (deprecated) 
	msg:addU16(0) -- bonus for lov level
	msg:addU16(0) -- XPBoost
	msg:addU16(100) -- 100% bonus of base exp bonus 
	msg:addU16(100) -- stamina boost(100=x1.0)
	msg:addByte(0x00) -- can buy exp boost

	msg:addU16(self:getHealth())
	msg:addU16(self:getMaxHealth())
	msg:addU16(self:getMana())
	msg:addU16(self:getMaxMana())
	msg:addByte(self:getSoul())
	msg:addU16(self:getStamina())

	local foodTime = 0
	local condition = self:getCondition(CONDITION_REGENERATION, CONDITIONID_DEFAULT)
	if condition then
		foodTime = condition:getTicks() /  1000
	end

	msg:addU16(foodTime) --health tick??
	msg:addU16(self:getOfflineTrainingTime() / 60 / 1000)
	msg:addU16(self:getSpeed() / 2)
	msg:addU16(self:getBaseSpeed() / 2)
	msg:addU32(self:getCapacity()) -- cap + boost
	msg:addU32(self:getCapacity())
	msg:addU32(self:getFreeCapacity())

	msg:addByte(8) -- ??
	msg:addByte(1) -- ??
	msg:addU16(self:getMagicLevel())
	msg:addU16(self:getBaseMagicLevel())
	-- loyalty bonus
	msg:addU16(self:getBaseMagicLevel())
	msg:addU16(self:getSkillPercent(SKILL_MAGLEVEL))

	for i = SKILL_FIST, SKILL_FISHING, 1 do
		msg:addByte(skillsId[i - 1])
		msg:addU16(self:getEffectiveSkillLevel(i))
		msg:addU16(self:getSkillLevel(i))
		-- loyalty bonus
		msg:addU16(self:getSkillLevel(i))
		msg:addU16(self:getSkillPercent(i) * 100)
	end

	msg:addByte(0)

	msg:sendToPlayer(self)
	msg:delete()
	return true
end

function Player.sendBadges(self, msg)
	msg:addByte(1) -- show account info?
	msg:addByte(1) -- account online?
	msg:addByte(self:isPremium() and 1 or 0) -- is premium?
	msg:addString("Loyalty Title") -- loyalty title

	msg:addByte(0) -- badges count
	-- structure:
	-- u32 badge id
	-- string badge name

	msg:sendToPlayer(self)
	msg:delete()
	return true
end

function Player.sendCombatStats(self, msg)
	for skillId = SPECIALSKILL_CRITICALHITCHANCE, SPECIALSKILL_MANALEECHAMOUNT do
		msg:addU16(self:getSpecialSkill(skillId))
		msg:addU16(0)
	end

	-- fatal, dodge, momentum
	for i = 1, 3 do
		msg:addU16(0)
		msg:addU16(0)
	end

	msg:addU16(0) -- cleave (bonus percent damage to nearby enemies)
	msg:addU16(0) -- bonus magic shield capacity (flat)
	msg:addU16(0) -- bonus magic shield capacity (percent)

	-- perfect shot flat damage bonus at 1-5 range
	for i = 1, 5 do
		msg:addU16(0)
	end

	msg:addU16(0) -- damage reflection (flat, one value for all combat types)
	msg:addByte(self:getBlessings()) -- blesssings used
	msg:addByte(8) -- blessings count

	-- weapon
	msg:addU16(0) -- base max damage
	msg:addByte(0) -- base element type
	msg:addByte(0) -- percent damage conversion
	msg:addByte(0) -- conversion element type

	msg:addU16(0) -- armor
	msg:addU16(0) -- defense

	-- element resistances count
	msg:addByte(0)
	-- structure:
	-- u8 clientcombat
	-- u8 value

	-- active potions count
	msg:addByte(0)
	-- structure:
	-- item clientId
	-- u16 duration

	msg:sendToPlayer(self)
	msg:delete()
	return true
end
```

**Comment by ArturKnopik** [comment]
I am currently closing this PR, I have a big problem how to get information about the wanda element damage, if I come up with something sensible, I will open it again

**Comment by EPuncker** [comment]
> I am currently closing this PR, I have a big problem how to get information about the wanda element damage, if I come up with something sensible, I will open it again

not everything should be working from start, just add comments to information that are not available yet, there is no need to be 100% working, just the basic structure is already welcome

**Pull request closed without merging**